modules/orchestrator/utils/pipeline_utils.go

```diff
@@ -134,24 +137,78 @@ func FindCreatingRuntimesByRelease(appID uint64, env string, ymlName string, bdl
-					ClusterName:  v.ClusterName,
-					Extra: map[string]interface{}{"applicationId": v.FilterLabels["appID"], "buildId": v.ID,
-						"workspace": v.Extra.DiceWorkspace, "commitId": v.Commit, "fakeRuntime": true},
-					TimeCreated: *v.TimeBegin,
-					CreatedAt:   *v.TimeBegin,
-					UpdatedAt:   *v.TimeBegin,
-				},
-				LastOperateTime: *v.TimeBegin,
-				LastOperator:    fmt.Sprintf("%v", v.Extra.RunUser.ID),
-			})
+			result = append(result, v)
+		}
+	}
+
+	return result, nil
+}
+
+// FindCreatingRuntimesByRelease find those runtimes created through the release
+func FindCreatingRuntimesByRelease(appID uint64, envs map[string][]string, ymlName string, bdl *bundle.Bundle) ([]apistructs.RuntimeSummaryDTO, error) {
+	var result []apistructs.RuntimeSummaryDTO
+	pipelinePageReq := apistructs.PipelinePageListRequest{
+		Sources:  []apistructs.PipelineSource{apistructs.PipelineSourceDice},
+		Statuses: []string{"Running"},
+	}
+	if appID != 0 {
+		pipelinePageReq.MustMatchLabelsQueryParams = []string{"appID=" + strconv.FormatUint(appID, 10)}
+	}
+	if ymlName != "" {
+		pipelinePageReq.YmlNames = []string{ymlName}
+	}
+
+	resp, err := bdl.PageListPipeline(pipelinePageReq)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range resp.Pipelines {
+		branchSlice := strings.SplitN(v.YmlName, "-", -1)
+		if len(branchSlice) != 4 {
+			return nil, errors.Errorf("Invalid yaml name %s", v.YmlName)
+		}
+		branch := branchSlice[3]
+		runtimeBranchs, ok := envs[strings.ToLower(v.Extra.DiceWorkspace)]
+		// first condition means user have permission
+		// second condition means the pipeline is used to deploy runtime by release
+		// third condition means that the runtime data of db has higher priority
+		// And one branch corresponds to only one rutime
+		if ok && strings.Contains(v.YmlName, "dice-deploy-release") && !strutil.Exist(runtimeBranchs, branch) {
+			// get pipeline detail to confirm whether the runtime has been created
+			piplineDetail, err := bdl.GetPipeline(v.ID)
+			if err != nil {
+				return nil, err
+			}
+			if isUndoneDeployByReleaseTask(piplineDetail) {
+				result = append(result, apistructs.RuntimeSummaryDTO{
+					RuntimeInspectDTO: apistructs.RuntimeInspectDTO{
+						Name:         v.FilterLabels["branch"],
+						Source:       apistructs.RELEASE,
+						Status:       "Init",
+						DeployStatus: apistructs.DeploymentStatusDeploying,
+						ClusterName:  v.ClusterName,
+						Extra: map[string]interface{}{"applicationId": v.FilterLabels["appID"], "buildId": v.ID,
+							"workspace": v.Extra.DiceWorkspace, "commitId": v.Commit, "fakeRuntime": true},
+						TimeCreated: *v.TimeBegin,
+						CreatedAt:   *v.TimeBegin,
+						UpdatedAt:   *v.TimeBegin,
+					},
+					LastOperateTime: *v.TimeBegin,
+					LastOperator:    fmt.Sprintf("%v", v.Extra.RunUser.ID),
+				})
+			}
 		}
 	}
 
 	return result, nil
 }
+
+func isUndoneDeployByReleaseTask(piplineDetail *apistructs.PipelineDetailDTO) bool {
```

Comment:
what does `UndoneDeploy` mean?

Reply:
The function name has been modified and commented